File: javascript/watch.js
/*
watch.js
    MediaWiki Action API Code Samples
    Demo of `Watch` module: Add a page to your watchlist
    MIT license
*/
var fetch = require('node-fetch');
var url = "https://en.wikipedia.org/w/api.php";
url = url + "?";
var cookie;

// Step 1: GET Request to fetch login token
function getLoginToken() {
    var PARAMS_1 = {
        action: "query",
        meta: "tokens",
        type: "login",
        format: "json"
    };

    var query = url;
    Object.keys(PARAMS_1).forEach(function (key) {
        query += "&" + key + "=" + PARAMS_1[key];
    });

    fetch(query)
        .then(function (response) {
            return response.json();
        })
        .then(function (data) {
            doLogin(data.query.tokens.logintoken);
        })
        .catch(function (error) {
            console.log(error);
        });
}
/*Step 2: Send a post request to log in. For this login method, Obtain credentials by first visiting https://www.en.wikipedia.org/wiki/Special:BotPasswords
See https://www.mediawiki.org/wiki/API:Login for more information on log in methods.*/
function doLogin(LOGIN_TOKEN){
	var PARAMS_2 = {
    action:"login",
    lgname:"your_bot_username",
    lgpassword:"your_bot_password",
    lgtoken:LOGIN_TOKEN,
    format:"json"
};
  fetch(url,{credentials:"include",
              method:"POST",
			  body:JSON.stringify(PARAMS_2),
			  header:{
				  "Content-type":"application/json"
			  }
  })
  .then(function(response){
	  var setCookie = response.headers.get('set-cookie');
      cookie = setCookie.substr(0, setCookie.indexOf(';'));
	  getCsrfToken();
  })
  .catch(function (error){
	  console.log(error);
  });
  }
  //retrieve a CSRF token
 function getCsrfToken(){
	 var PARAMS_3 = {
    action: "query",
    meta: "tokens",
    type: "watch",
    format: "json"
};
  var query = url;
    Object.keys(PARAMS_3).forEach(function (key) {
        query += "&" + key + "=" + PARAMS_3[key];
    });
	fetch(query, {
            credentials: "include",
            headers: {
                "Content-Type": "application/json",
                "Cookie": cookie,
            }
        })
        .then(function (response) {
            return response.json();
        })
        .then(function (data) {
            addPage(data.query.tokens.watchtoken);
        })
        .catch(function (error) {
            console.log(error);
        });
 }
 //Post request to add a page to your watchlist
 function addPage(CSRF_TOKEN){
	 var PARAMS_4 = {
    action: "watch",
    titles: "Stone forest",
    format: "json",
    token: CSRF_TOKEN,
}
fetch(url,{credentials: "include",
            method: "POST",
            body: JSON.stringify(PARAMS_4),
            headers: {
                "Content-Type": "application/json",
                "Cookie": cookie,
            }
        })
        .then(function (response) {
            return response.json();
        })
        .then(function (data) {
            console.log(data);
        })
        .catch(function (error) {
            console.log(error);
        });
 }

File: python/watch.py
#!/usr/bin/python3

"""
    watch.py

    MediaWiki API Demos
    Demo of `Watch` module: Add a page to your watchlist
    MIT license
"""

import requests

S = requests.Session()

URL = "https://en.wikipedia.org/w/api.php"

# Step 1: Retrieve a login token
PARAMS_1 = {
    "action": "query",
    "meta": "tokens",
    "type": "login",
    "format": "json"
}

R = S.get(url=URL, params=PARAMS_1)
DATA = R.json()

LOGIN_TOKEN = DATA["query"]["tokens"]["logintoken"]

# Step 2: Send a post request to log in. For this login
# method, Obtain credentials by first visiting
# https://www.en.wikipedia.org/wiki/Special:BotPasswords
# See https://www.mediawiki.org/wiki/API:Login for more
# information on log in methods.
PARAMS_2 = {
    "action": "login",
    "lgname": "username",
    "lgpassword": "password",
    "format": "json",
    "lgtoken": LOGIN_TOKEN
}

R = S.post(URL, data=PARAMS_2)

# Step 3: While logged in, retrieve a CSRF token
PARAMS_3 = {
    "action": "query",
    "meta": "tokens",
    "type": "watch",
    "format": "json"
}

R = S.get(url=URL, params=PARAMS_3)
DATA = R.json()

CSRF_TOKEN = DATA["query"]["tokens"]["watchtoken"]

# Step 4: Post request to add a page to your watchlist
PARAMS_4 = {
    "action": "watch",
    "titles": "Stone forest",
    "format": "json",
    "token": CSRF_TOKEN,
}

R = S.post(URL, data=PARAMS_4)
DATA = R.json()

print(DATA)
